File: app/src/main/java/com/moisesborges/tvaddict/shows/ShowsFragment.java
package com.moisesborges.tvaddict.shows;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.moisesborges.tvaddict.App;
import com.moisesborges.tvaddict.R;
import com.moisesborges.tvaddict.adapters.ItemClickListener;
import com.moisesborges.tvaddict.adapters.ShowsAdapter;
import com.moisesborges.tvaddict.models.Show;
import com.moisesborges.tvaddict.showdetails.ShowDetailsActivity;
import com.moisesborges.tvaddict.utils.ProgressBarHelper;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import timber.log.Timber;

/**
 * Created by Moisés on 12/04/2017.
 */

public class ShowsFragment extends Fragment implements ShowsView {

    private Unbinder mUnbinder;

    @BindView(R.id.shows_recycler_view)
    RecyclerView mShowsRecyclerView;
    @BindView(R.id.progress_bar)
    ProgressBar mProgressBar;

    @Inject
    ShowsPresenter mShowsPresenter;

    private ItemClickListener<Show> mShowItemClickListener = show -> mShowsPresenter.openShowDetails(show);

    private final ShowsAdapter mShowsAdapter = new ShowsAdapter(mShowItemClickListener);

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        ((App) context.getApplicationContext()).getAppComponent().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View layout = inflater.inflate(R.layout.fragment_shows, container, false);
        setRetainInstance(true);
        mUnbinder = ButterKnife.bind(this, layout);
        setupRecyclerView();
        return layout;
    }

    @Override
    public void onStart() {
        super.onStart();
        Timber.d("onStart");
        mShowsPresenter.bindView(this);
        mShowsPresenter.loadShows();
    }

    @Override
    public void onStop() {
        super.onStop();
        mShowsPresenter.unbindView();
    }

    private void setupRecyclerView() {
        mShowsRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), 3));
        mShowsRecyclerView.setAdapter(mShowsAdapter);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
    }

    @Override
    public void displayProgress(boolean isLoading) {
        ProgressBarHelper.show(mProgressBar, isLoading);
    }

    @Override
    public void displayTvShows(@NonNull List<Show> shows) {
        mShowsAdapter.setShows(shows);
    }

    @Override
    public void displayError() {

    }

    @Override
    public void navigateToShowDetails(@NonNull Show show) {
        ShowDetailsActivity.start(getContext(), show);
    }

    public static Fragment newInstance() {
        return new ShowsFragment();
    }

}

File: app/src/main/java/com/moisesborges/tvaddict/shows/ShowsFragment.kt
package com.moisesborges.tvaddict.shows

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.GridLayoutManager
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.moisesborges.tvaddict.App
import com.moisesborges.tvaddict.R
import com.moisesborges.tvaddict.adapters.ItemClickListener
import com.moisesborges.tvaddict.adapters.ShowsAdapter
import com.moisesborges.tvaddict.models.Show
import com.moisesborges.tvaddict.showdetails.ShowDetailsActivity
import com.moisesborges.tvaddict.ui.SpacesItemDecoration
import com.moisesborges.tvaddict.utils.ProgressBarHelper
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.fragment_shows.*
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by Moisés on 12/04/2017.
 */

class ShowsFragment : Fragment(), ShowsView {

    @Inject
    internal lateinit var showsPresenter: ShowsPresenter

    private val showItemClickListener: ItemClickListener<Show> = ItemClickListener { show -> showsPresenter.openShowDetails(show) }
    private val followShowListener = ItemClickListener<Show> { show -> showsPresenter.followShow(show)}

    private val showsAdapter = ShowsAdapter(showItemClickListener, followShowListener)

    private val compositeDisposable = CompositeDisposable()

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        (context!!.applicationContext as App).appComponent.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val layout = inflater!!.inflate(R.layout.fragment_shows, container, false)
        retainInstance = true
        return layout
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupRecyclerView()
    }

    override fun onStart() {
        super.onStart()
        Timber.d("onStart")
        showsPresenter.bindView(this)
        showsPresenter.loadShows()
    }

    override fun onResume() {
        super.onResume()
        compositeDisposable.add(listenToPageEvents())
    }

    private fun listenToPageEvents(): Disposable {
        return showsAdapter.lastItemReachedEvent()
                .subscribe({ currentPage -> showsPresenter.loadMoreShows(currentPage) },
                        { error -> Timber.e(error) })
    }

    override fun onPause() {
        super.onPause()
        compositeDisposable.clear()
    }

    override fun onStop() {
        super.onStop()
        showsPresenter.unbindView()
    }

    private fun setupRecyclerView() {
        val rows = 3
        shows_recycler_view.layoutManager = GridLayoutManager(context, rows)
        shows_recycler_view.addItemDecoration(SpacesItemDecoration(16, rows))
        shows_recycler_view.adapter = showsAdapter
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    override fun displayProgress(isLoading: Boolean) {
        ProgressBarHelper.show(progress_bar, isLoading)
    }

    override fun displayTvShows(shows: List<Show>) {
        showsAdapter.setShows(shows)
    }

    override fun displayError() {

    }

    override fun navigateToShowDetails(show: Show) {
        ShowDetailsActivity.start(context, show)
    }

    override fun setPage(page: Int) {
        showsAdapter.setPage(page)
    }

    override fun displayMoreTvShows(shows: List<Show>) {
        showsAdapter.addShows(shows)
    }

    override fun displayFollowingShowMessage(show: Show) {
        val text = context.resources.getString(R.string.following_show_message, show.name)
        Toast.makeText(context, text, Toast.LENGTH_SHORT).show()
    }

    override fun hideShow(show: Show) {
        showsAdapter.removeShow(show)
    }

    companion object {

        @JvmStatic
        fun newInstance(): Fragment {
            return ShowsFragment()
        }
    }

}

